lab/skel_graph/shortest_path/johnson/main.cpp

```diff
@@ -0,0 +1,226 @@
+// Copyright 2020 Miruna-Elena Banu <mirunaelena.banu@gmail.com>
+
+// Gasirea celui mai scurt drum intre oricare doua noduri
+// All-Pairs Shortest Path Problem
+
+// Johnson - T = O(n * m  * log n + n * m) (Binary Heap)
+//         - T = O(n^2 * log n + n * m) (Fibonacci Heap)
+
+// links:
+// - TODO: Gaseste un link mai bun
+// - https://infoarena.ro/problema/royfloyd
+
+#include <bits/stdc++.h>
+
+#define NMAX      50010     // ATENTIE la cat e in problema curenta !!!
+#define oo        (1 << 30) // ATENTIE la cat e in problema curenta !!!
+#define NO_PARENT (-1)
+
+using namespace std;
+
+class Task {
+public:
+    void solve() {
+        read_input();
+        print_output(get_result());
+    }
+
+private:
+    // n = numar de noduri, m = numar de muchii
+    int n, m;
+
+    // Listele de adiacenta pentru fiecare nod
+    int adj[NMAX][NMAX];
+
+    void read_input() {
+        cin >> n;
+        for (int x = 1; x <= n; ++x) {
+            for (int y = 1; y <= n; ++y) {
+                int c; // x -> y, of cost c
+                cin >> c;
+                if (c == 0) {
+                    c = oo;
+                }
+                adj[x][y] = c;
+            }
+        }
+    }
+
+    bool get_result() { return Johnson(); }
+
+    void Dijkstra(int source, vector<int>& d) {
+        // Min-heap ce contine perechi de tipul: distanta source -> x
+        // Folosit pentru a extrage la fiecare pas nodul care are costul
+        // estimat minim fata de sursa
+        priority_queue<pair<int, int>, vector<pair<int, int>>, std::greater<pair<int, int>>> pq;
+
+        // Presupunem ca nu exista drum de la source la i
+        for (int i = 1; i <= n; ++i) {
+            d[i] = oo;
+        }
+
+        // Initializam distanta source -> source = 0
+        d[source] = 0;
+        // Adaugam in heap sursa cu prioritate = cost
+        pq.push({d[source], source});
+
+        while (!pq.empty()) {
+            // Extragem nodul cu costul estimat minim fata de sursa
+            auto entry = pq.top();
+            pq.pop();
+
+            int cost = entry.first;
+            int node = entry.second;
```

Comment:
Let's use `-std=c++17` in Makefile.

we can say `auto [cost, node] = pq.top();`